### ISSUE TITLE
Error: not found: value LivyUtils

### ISSUE THREAD

**Issue by markuszinser** [sparklyr/sparklyr#2616]
We are currently trying to setup a yarn spark cluster with livy. We are running spark 3.0.0 with scala 2.12. The configuration seems to work when running pyspark. But when using sparklyr we are getting some errors. Somehow the necessary jars for sparklyr can't be downloaded from github (proxy on the master is usually working). This is why I put all of them on the hdfs and tried to tell sparklyr to collect them from there. This seems to work but we are now getting the error that the LivyUtils can't be loaded. 

Script to connect:
```r
library(sparklyr)
conf <- spark_config()
conf$sparklyr.livy.sources <- TRUE
config <- livy_config(
    config = conf,
    jars = "hdfs:///tmp/sparklyr-*.jar"
)
sc <- spark_connect(
    master = "http://***.***.***.***:8998",
    method = "livy",
    config = config,
    version = "3.0.0",
    scala_version = "2.12"
)
```

The following error is thrown when trying to connect:
```
Error in h(simpleError(msg, call)) : 
  Failed to initialize livy connection: Failed to execute Livy statement with error: <console>:24: error: not found: value LivyUtils
Traceback:        var sparklyrRetVar_1 = LivyUtils.invokeFromBase64("AAAAHm9yZy5hcGFjaGUuc3BhcmsuU3BhcmtDb250ZXh0AAAAAAEAAAAMZ2V0T3JDcmVhdGUAAAAAAA==")
                              ^


callstack:
spark_connect(master = "http://***.***.***.***:8998", method = "livy", config = config, version = "3.0.0", scala_version = "2.12")
initialize_connection(scon)
initialize_connection.livy_connection(scon)
withCallingHandlers({
    if (spark_config_value(sc$config, "sparklyr.livy.sources", TRUE)) {
        livy_load_scala_sources(sc)
    }
    session <- tryCatch({
        invoke_static(sc, "org.apache.spark.sql.SparkSession", "builder") %>% invoke("%>%", list("enableHiveSupport"), list("getOrCreate"))
    }, error = function(e) {
        NULL
    })
    sc$state$spark_context <- tryCatch({
        invoke(session, "sparkContext")
    }, error = function(e) {
        invoke_static(sc, "org.apache.spark.SparkContext", "getOrCreate")
    })
    sc$state$java_context <- invoke_static(sc, "org.apache.spark.api.java.JavaSparkContext", "fromSparkContext", spark_context(sc))
    sc$state$spark_version <- spark_version(sc)
    sc$state$hive_context <- session %||% create_hive_context(sc)
    if (spark_version(sc) < "2.0.0") {
        params <- connection_config(sc, "spark.sql.")
        apply_config(hive_context, params, "setConf", "spark.sql.")
    }
    sc
}, error = function(err) {
    stop("Failed to initialize livy connection: ", err$message, "\n\ncallstack:\n", paste(sys.calls(), collapse = "\n"))
})
tryCatch({
    invoke(session, "sparkContext")
}, error = function(e) {
    invoke_static(sc, "org.apache.spark.SparkContext", "getOrCreate")
})
tryCatchList(expr, classes, parentenv, handlers)
tryCatchOne(expr, names, parentenv, handlers[[1]])
value[[3]](cond)
invoke_static(sc, "org.apache.spark.SparkContext", "getOrCreate")
invoke_static.livy_connection(sc, "org.apache.spark.SparkContext", "getOrCreate")
livy_invoke_statement_fetch(sc, TRUE, classMapped, method, ...)
livy_invoke_statement(sc, statement)
livy_post_statement(sc, statement$code)
withr::with_options(list(warning.length = 8000), {
    stop("Failed to execute Livy statement with error: ", if (is.null(statementReponse$output$evalue)) 
        jsonlite::toJSON(statementReponse)
    else statementReponse$output$evalue, "\nTraceback: ", paste(statementReponse$output$traceback, collapse = ""))
})
force(code)
stop("Failed to execute Livy statement with error: ", if (is.null(statementReponse$output$evalue)) jsonlite::toJSON(statementReponse) else statementReponse$output$evalue, "\nTraceback: ", paste(statementReponse$output$traceback, collapse = ""))
.handleSimpleError(function (err) 
{
    stop("Failed to initialize livy connection: ", err$message, "\n\ncallstack:\n", paste(sys.calls(), collapse = "\n"))
}, "Failed to execute Livy statement with error: <console>:24: error: not found: value LivyUtils\nTraceback:        var sparklyrRetVar_1 = LivyUtils.invokeFromBase64(\"AAAAHm9yZy5hcGFjaGUuc3BhcmsuU3BhcmtDb250ZXh0AAAAAAEAAAAMZ2V0T3JDcmVhdGUAAAAAAA==\")\n                              ^\n", base::quote(force(code)))
h(simpleError(msg, call))
```

Does someone have any idea what is happening? Is livy only supported for spark 2.3? I saw that only
until spark 2.3 the LivyUtils class is present. Or is this also compiled into the jars for 3.0.0?

**Comment by yl790** [comment]
@markuszinser Hey I'll look into this ASAP.

I can confirm the `LivyUtils` class is definitely compiled in sparklyr-3.0-2.12.jar, so I don't think that's the issue here. You can double-check this is the case by running `jar tf <path to your sparklyr installation>/java/sparklyr-3.0-2.12.jar` and see `LivyUtils.class` is in the output.

**Comment by yl790** [comment]
@markuszinser My best guess is to try the following config instead:

```
conf <- spark_config()
conf$sparklyr.livy.sources <- TRUE
conf$sparklyr.livy.jar <- "hdfs:///tmp/sparklyr-3.0-2.12.jar"
config <- livy_config(
    config = conf,
    jars = "hdfs:///tmp/sparklyr-3.0-2.12.jar"  # actually I don't think you need this... also, I strongly suspect
                                                                     # the '*' expansion will not work here like it would work in
                                                                     # bash (but could be wrong)
)
```

**Comment by markuszinser** [comment]
@yl790 Thanks a lot for the reply. Indeed, the LivyUtils are included.

I changed the config to what you suggested, but it unfortunately doesn't help. 

In the meantime I realized that no matter what path I enter, I always get the same error. Could it be that sparklyr is looking at the wrong place and doesn't find the jar? The file definitely exists. And I have the feeling that it also is used, because when I accidentally put a broken jar into the directory I got a nullptr exception. I also tried to use a local directory and added it to the livy whitelist. But I still get the same directory.
